config/archive.go

```diff
@@ -0,0 +1,73 @@
+package config
+
+import (
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/contiv/volplugin/errors"
+	"github.com/contiv/volplugin/watch"
+	"github.com/coreos/etcd/client"
+	"golang.org/x/net/context"
+)
+
+func (c *Client) policyArchive(name string) string {
+	return c.prefixed(rootPolicyArchive, name)
+}
+
+func (c *Client) policyArchiveEntry(name string, timestamp string) string {
+	return c.prefixed(rootPolicyArchive, name, timestamp)
+}
+
+// CreatePolicyRevision creates an revision entry in a policy's history.
+func (c *Client) CreatePolicyRevision(name string, policy string) error {
+	timestamp := fmt.Sprint(time.Now().Unix())
+	key := c.policyArchiveEntry(name, timestamp)
+
+	_, err := c.etcdClient.Set(context.Background(), key, policy, nil)
+	if err != nil {
+		return errors.EtcdToErrored(err)
+	}
+
+	return nil
+}
+
+// ListPolicyRevisions returns a list of all the revisions for a given policy.
+func (c *Client) ListPolicyRevisions(name string) ([]string, error) {
+	keyspace := c.policyArchive(name)
+
+	resp, err := c.etcdClient.Get(context.Background(), keyspace, &client.GetOptions{Sort: true, Recursive: false, Quorum: true})
+	if err != nil {
+		return nil, errors.EtcdToErrored(err)
+	}
+
+	ret := []string{}
+
+	for _, node := range resp.Node.Nodes {
+		_, revision := path.Split(node.Key)
+
+		ret = append(ret, revision)
+	}
+
+	return ret, nil
+}
+
+// GetPolicyRevision returns a single revision for a given policy.
+func (c *Client) GetPolicyRevision(name, revision string) (string, error) {
+	keyspace := c.policyArchiveEntry(name, revision)
+
+	resp, err := c.etcdClient.Get(context.Background(), keyspace, &client.GetOptions{Sort: false, Recursive: false, Quorum: true})
+	if err != nil {
+		return "", errors.EtcdToErrored(err)
+	}
+
+	return resp.Node.Value, nil
+}
+
+// WatchForPolicyChanges creates a watch which prints the policy name and revision number
+// (timestamp) to stdout whenever a policy is uploaded.
+func (c *Client) WatchForPolicyChanges(ch chan *watch.Watch, f func(resp *client.Response, w *watch.Watcher)) {
+	w := watch.NewWatcher(ch, c.prefixed(rootPolicyArchive), f)
```

Comment:
it would be nice if this did more than provide a facade for the watch lib. Perhaps it should unmarshal the policies, or otherwise fill the `Config` variable.

sorry I missed this on the first pass. I just think watch is being misused here. This is more of what I had in mind for this library: https://github.com/contiv/volplugin/blob/master/config/volume.go#L292-L318